src/tools/strings.ml

```diff
@@ -1,144 +1,275 @@
 (*****************************************************************************
-
+  
   Liquidsoap, a programmable audio stream generator.
   Copyright 2003-2019 Savonet team
-
+  
   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.
-
+  
   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details, fully stated in the COPYING
   file at the root of the liquidsoap distribution.
-
+  
   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA
-
+  
  *****************************************************************************)
-
+  
 (** Operations on lists of strings. This is module is used in order to avoid
     concatenating (large) strings. Iterators are FIFO. *)
 
+module type Strings = sig
+  (** A buffer of strings. *)
+  type buffer
+
+  type return
+
+  (* Create a buffer from a string. *)
+  val of_string : string -> buffer
+
+  (* Create a buffer from the given bytes.
+   * bytes will be copied. *)
+  val of_bytes : bytes -> buffer
+
+  (* Create a buffer from the given bytes.
+   * bytes will not be copied. *)
+  val unsafe_of_bytes : bytes -> buffer
+
+  (** Render a buffer into a string. This operation can be costly (in terms of memory copies), avoid it. *)
+  val to_string : buffer -> string
+
+  val substring : buffer -> int -> int -> string
+
+  (** Concatenation of strings. *)
+  val of_list : string list -> buffer
+
+  (** Add a string at the end of a buffer. *)
+  val add : buffer -> string -> return
+
+  (** Add bytes at the end of a buffer.
+    * bytes will be copied. *)
+  val add_bytes : buffer -> bytes -> return
+
+  (** Add bytes at the end of a buffer.
+    * bytes will not be copied. *)
+  val unsafe_add_bytes : buffer -> bytes -> return
```

Comment:
I would rather not have such functions in the API and let the user do the `Bytes.unsafe_to_string` by himself: we want to make it hard for everybody to do unsafe casts since they are potentially quite unsafe (e.g. see my comments on your other changes).

Reply:
I think any pattern that leads to repeated nested call should be abstracted away in a API call to make the code more easily readable. Let's see what's left of this API after we commit the fixes you spotted.